pkg/storage/mvcc.go

```diff
@@ -5153,6 +5153,80 @@ func MVCCGarbageCollectRangeKeys(
 	return nil
 }
 
+// MVCCGarbageCollectWholeRange removes all the range data and resets counters.
+// It only does so if data is completely covered by range keys
+func MVCCGarbageCollectWholeRange(
+	ctx context.Context,
+	rw ReadWriter,
+	ms *enginepb.MVCCStats,
+	start, end roachpb.Key,
+	gcThreshold hlc.Timestamp,
+) error {
+	if ms.ContainsEstimates == 0 && ms.LiveCount > 0 {
+		return errors.Errorf("range contains live data, can't use GC clear range")
+	}
+	if err := CheckRangeEmptiness(ctx, rw, start, end, gcThreshold); err != nil {
+		return err
+	}
+	if err := rw.ClearRawRange(start, end, true, true); err != nil {
+		return err
+	}
+	if ms != nil {
+		// Reset point and range counters as we deleted the whole range.
+		ms.LiveCount = 0
+		ms.LiveBytes = 0
+		ms.KeyCount = 0
+		ms.KeyBytes = 0
+		ms.ValCount = 0
+		ms.ValBytes = 0
+		ms.RangeKeyCount = 0
+		ms.RangeKeyBytes = 0
+		ms.RangeValCount = 0
+		ms.RangeValBytes = 0
+	}
+	return nil
+}
+
+// CheckRangeEmptiness checks if a span of keys doesn't contain any live data
+// and all data is below provided threshold. This functions is meant for fast
+// path deletion by GC where range can be removed by a range tombstone.
+func CheckRangeEmptiness(
+	ctx context.Context, rw Reader, start, end roachpb.Key, gcThreshold hlc.Timestamp,
+) error {
+	// It makes no sense to check local ranges for fast path.
+	if isLocal(start) || isLocal(end) {
```

Comment:
I wonder if we should exclude system data from this fast path too, just to be safe. I.e. anything below `SystemMax`.